apm-agent-plugins/apm-vertx/apm-vertx3-plugin/src/main/java/co/elastic/apm/agent/vertx/v3/web/HttpServerRequestImplEndInstrumentation.java

```diff
@@ -16,47 +16,63 @@
  * specific language governing permissions and limitations
  * under the License.
  */
-package co.elastic.apm.agent.vertx.v3.web.http1;
+package co.elastic.apm.agent.vertx.v3.web;
 
-import co.elastic.apm.agent.vertx.v3.web.WebHelper;
-import co.elastic.apm.agent.vertx.v3.web.WebInstrumentation;
+import co.elastic.apm.agent.impl.transaction.Transaction;
+import co.elastic.apm.agent.sdk.logging.Logger;
+import co.elastic.apm.agent.sdk.logging.LoggerFactory;
+import io.vertx.core.http.HttpServerRequest;
+import io.vertx.core.http.HttpServerResponse;
 import io.vertx.core.http.impl.HttpServerRequestImpl;
 import net.bytebuddy.asm.Advice;
 import net.bytebuddy.description.method.MethodDescription;
 import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.implementation.bytecode.assign.Assigner;
 import net.bytebuddy.matcher.ElementMatcher;
 
+import javax.annotation.Nullable;
+
 import static net.bytebuddy.matcher.ElementMatchers.named;
-import static net.bytebuddy.matcher.ElementMatchers.takesNoArguments;
 
 /**
- * Instruments {@link io.vertx.core.http.impl.HttpServerRequestImpl#doEnd()} to remove the context from the context map again.
+ * Instruments {@link HttpServerRequestImpl#handleEnd()} and {@link io.vertx.core.http.impl.Http2ServerRequestImpl#handleEnd} to finalize
+ * the transaction and remove the mapping.
  */
 @SuppressWarnings("JavadocReference")
 public class HttpServerRequestImplEndInstrumentation extends WebInstrumentation {
 
     @Override
     public ElementMatcher<? super TypeDescription> getTypeMatcher() {
-        return named("io.vertx.core.http.impl.HttpServerRequestImpl");
+        return named("io.vertx.core.http.impl.HttpServerRequestImpl")
+            .or(named("io.vertx.core.http.impl.Http2ServerRequestImpl"));
     }
 
     @Override
     public ElementMatcher<? super MethodDescription> getMethodMatcher() {
-        return named("doEnd").or(named("onEnd")).and(takesNoArguments());
+        return named("handleEnd");
     }
 
     @Override
     public String getAdviceClassName() {
-        return "co.elastic.apm.agent.vertx.v3.web.http1.HttpServerRequestImplEndInstrumentation$HttpRequestEndAdvice";
+        return "co.elastic.apm.agent.vertx.v3.web.HttpServerRequestImplEndInstrumentation$HttpRequestEndAdvice";
     }
 
     public static class HttpRequestEndAdvice {
 
+        private static final Logger log = LoggerFactory.getLogger(HttpRequestEndAdvice.class);
+
         private static final WebHelper helper = WebHelper.getInstance();
 
         @Advice.OnMethodExit(suppress = Throwable.class, inline = false)
-        public static void exit(@Advice.This HttpServerRequestImpl request) {
-            helper.removeTransactionFromContext(request);
+        public static void exit(@Advice.This HttpServerRequest request,
+                                @Advice.FieldValue(value = "response", typing = Assigner.Typing.DYNAMIC) @Nullable HttpServerResponse response) {
```

Comment:
We are using this Advice to instrument bot, `HttpServerRequestImpl` and `Http2ServerRequestImpl` (see type matcher above), BUT, here you use `HttpServerResponse` which is only valid for `HttpServerRequestImpl`. Response class for `Http2ServerRequestImpl` is called `Http2ServerResponseImpl`. So, with Http2 we'll never finalize the transaction with the actual response information with this instrumentation, right? So we may want to split this into a separate advice or do some casting of the response object

Reply:
`io.vertx.core.http.HttpServerResponse` is the interface of both and I specified I allow dynamic typing for the inference of this field